pkg/pf/tests/internal/providerbuilder/build_resource.go

```diff
@@ -17,10 +17,61 @@ package providerbuilder
 import (
 	"context"
 
+	"github.com/hashicorp/terraform-plugin-framework/path"
 	"github.com/hashicorp/terraform-plugin-framework/resource"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
 )
 
+type NewResourceArgs struct {
+	// Name is the name of the resource. Defaults to "test".
+	Name           string
+	ResourceSchema schema.Schema
+
+	CreateFunc      func(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse)
+	ReadFunc        func(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse)
+	UpdateFunc      func(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse)
+	DeleteFunc      func(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse)
+	ImportStateFunc func(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse)
+}
+
+// NewResource creates a new Resource with the given parameters, filling reasonable defaults.
+func NewResource(args NewResourceArgs) Resource {
```

Comment:
It's weird to me that it's valid to create a `pb.Resource` and set fields directly *or* with `pb.NewResource(pb.NewResourceArgs{})`. Since `pb.Resource` and `pb.ResourceArgs` are **exactly** the same struct, we should create a "correct" way to get a `pb.Resource`. 

IMO we should:

1) Give `pb.Resource` a private embedding of `pb.ResourceArgs`, making it impossible to create a `pb.Resource` without calling `pb.NewResource`. 

or

2) Remove `pb.NewResourceArgs` and change the name of `pb.NewResource` to `pb.ResourceWithDefaults(Resource) Resource` (which is what it actually does).

Reply:
I don't believe there's anything weird about this. This is a test utility, so it should be quite flexible to allow for various forms of tests. I've left a note in pb.Resource about NewResource. That is the way to get a valid Resource.

I think it's quite idiomatic in go for the constructor function to be called New or NewResources. You are correct that it is applying defaults but I don't think `pb.ResourceWithDefaults(Resource) Resource` is clearer.